src/Pos/Delegation/DB.hs

```diff
@@ -1,101 +1,239 @@
-import           Pos.DB.Class         (MonadDB, MonadDBPure)
-import           Pos.DB.Functions     (RocksBatchOp (..), encodeWithKeyPrefix)
-import           Pos.DB.GState.Common (gsGetBi)
-import           Pos.DB.Iterator      (DBIteratorClass (..), DBnIterator, DBnMapIterator,
-                                       IterType, runDBnIterator, runDBnMapIterator)
-import           Pos.DB.Types         (NodeDBs (_gStateDB))
-import           Pos.Types            (ProxySKHeavy, StakeholderId, addressHash)
-import           Pos.Util.Iterator    (nextItem)
+import           Control.Lens           (uses, (%=))
+import qualified Data.HashMap.Strict    as HM
+import qualified Data.HashSet           as HS
+import qualified Database.RocksDB       as Rocks
+
+import           Pos.Binary.Class       (encodeStrict)
+import           Pos.Crypto             (PublicKey, pskIssuerPk)
+import           Pos.DB.Class           (MonadDB, MonadDBPure)
+import           Pos.DB.Error           (DBError (DBMalformed))
+import           Pos.DB.Functions       (RocksBatchOp (..), encodeWithKeyPrefix)
+import           Pos.DB.GState.Common   (gsGetBi)
+import           Pos.DB.Iterator        (DBIteratorClass (..), DBnIterator,
+                                         DBnMapIterator, IterType, runDBnIterator,
+                                         runDBnMapIterator)
+import           Pos.DB.Types           (NodeDBs (_gStateDB))
+import           Pos.Delegation.Helpers (isRevokePsk)
+import           Pos.Delegation.Types   (DlgMemPool)
+import           Pos.Types              (ProxySKHeavy, StakeholderId, addressHash)
+import           Pos.Util.Iterator      (nextItem)
 
 ----------------------------------------------------------------------------
 -- Getters/direct accessors
 ----------------------------------------------------------------------------
 
--- | Retrieves certificate by issuer address (hash of public key) if present.
-getPSKByIssuerAddressHash :: MonadDBPure m => StakeholderId -> m (Maybe ProxySKHeavy)
-getPSKByIssuerAddressHash addrHash = gsGetBi (pskKey addrHash)
+-- | Retrieves certificate by issuer public key or his
+-- address/stakeholder id, if present.
+getPskByIssuer
+    :: MonadDBPure m
+    => Either PublicKey StakeholderId -> m (Maybe ProxySKHeavy)
+getPskByIssuer (either addressHash identity -> issuer) = gsGetBi (pskKey issuer)
+
+-- | Given an issuer, retrieves all certificate chains starting in
+-- issuer. This function performs a series of consequental db reads so
+-- it must be used under the shared lock.
+getPskChain
+    :: MonadDBPure m
+    => Either PublicKey StakeholderId -> m DlgMemPool
+getPskChain = getPskChainInternal HS.empty
+
+-- See doc for 'getPskTree'. This function also stops traversal if
+-- encounters anyone in 'toIgnore' set.
+getPskChainInternal
+    :: MonadDBPure m
+    => HashSet StakeholderId -> Either PublicKey StakeholderId -> m DlgMemPool
+getPskChainInternal toIgnore (either addressHash identity -> issuer) =
```

Comment:
seems, this function is wrong: you perform only one iteration, remove one element from list and break, because you don't add any elements to `_2`